proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,200 @@
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png",
+  },
+  {
+    "16": "dark16.png",
+    "32": "dark32.png",
+    "color_scheme": "dark",
+  },
+  {
+    "16": "light16.png",
+    "32": "light32.png",
+    "color_scheme": ["dark", "light"]
+  }
+]
+```
+
+Set icon_variants dynamically.
+```
+const exampleProperties: {string: string | ImageData}[] = [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png"
+  },
+  {
+    "16": darkImageData16,
+    "32": darkImageData32,
+    "color_scheme": "dark"
+  },
+  {
+    "16": lightImageData16,
+    "32": lightImageData32,
+    "color_scheme": "light"
+  }
+];
+
+// action.setIcon().
+const createProperties = {variants: exampleProperties};
+action.setIcon(createProperties);
+
+// menus.*(), for supporting browsers.
+const menusProperties = {icon_variants: exampleProperties};
+menus.create(menusProperties);
+menus.update(menusProperties);
+```
+
+A benefit of this new structure is that it's more resiliant to future changes, thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
+monochrome), and etc.
+
+## setIcon()
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+The `dark` icon variant is used when a dark color scheme is used for the
+surrounding UI, often referred to as "dark mode".
+The `light` icon variant is used otherwise.
+
+### New Permissions
+
+N/A.
+
+### Manifest File Changes
+
+1. If icon_variants are supplied, icons can be ignored.
+1. If icon_variants contain an icon object with duplicate conditions, the first match will be
+used.
+1. Any icon object that does not contain a "color_scheme" key will apply to both
+light and dark.
+1. If only one icon object is supplied, it will be used for both light and dark.
+1. icon_variants will not cause an error in the event that it is invalid
+Instead, it can be ignored altogether or just emit a warning. Warnings are
+preferred over errors because they're more adaptable to changes in the future.
+1. Neither "dark" nor "light" color_scheme's are required.
+1. Icon objects missing the color_scheme property will apply to light and dark
+mode. It could also be explicity set using {"color_scheme": ["dark", "light"]}.
+1. If icon_variants are provided, the top level icons key and
+action.default_icon will be ignored. There is no fallback.
+1. 16 is a size in `{"16": "icon.png"}` and any number as a size can be
+used, as per browser specifications. The word `"any"` could also be used in
+place of a number to represent a path to an icon with any supported format.
+
```

Comment:
To make user extension developers can rely on deterministic behaviour I suggest we specify how icon size fallbacks will be handled by browsers:
```suggestion
The icon size used by the browser will be determined as follows:
For raster images. The requested size will be used as defined in icon_variants. If the requested icon size by the browser is not present in the manifest file, a size higher than the requested icon size will be used if present.
For vector images. The size lookup will be density independent.
Example: the browser wants to use an icon with size 16 with a density 1.5. For raster images, it will search for an icon with size 24. While for vector images it will search for an icon with size 16.
```
This specific behaviour for vector images allows extensions to better control the amount of details they want to display in icons.

Reply:
I feel like this might be over-specifying, but interested to hear from other browser vendors.

Reply:
I think it is worth writing this down, even though it mostly follows the old rules. We should specify the unspecified from before. Plus enough has changed with the addition of `any` and SVG.

Reply:
@xeenon Agreed. We should probably also specify what should happen when both explicitly size (16, 32) and "any" are defined in the manifest. Should this be valid? I see a situation in which a browser would use the exact size if present, otherwise fall back to "any". How do you feel about that?

When it comes to vector size matching. If we want to go all in there are a few modes on which matching could take place.
1) DENSITY closest size * density. This is how matching would work for raster images with the exception that it is not rounded up as a closest match would not lead to side effects (blurry images).
2) EXACT closest size density independent.
3) Most specific dimensions based on integer ratios. This one is a bit more complicated. To reach pixel perfect icons, authors should specify the viewBox dimensions as size. The browser will then attempt to find the closest image size resulting in crispy images. Example: The browser wants a 48px icon. If the extension defines both 32px and 24px sizes in the manifest. The browser will pick 24px as this will make sure the icon is as pixel perfect as possible.

The mode used could potentially be explicitly set in the icon variants image group. Something like:
```json
[{
  "24": "path",
  "32": "path",
  "sizeMatching": ["ratio", "exact"]
}]
```

Reply:
At present, there is no [extensions specification](https://w3c.github.io/webextensions/specification/) for `icons` at all. But browsers are already very good at this. For example, there is "[preferred density-corrected dimensions](https://html.spec.whatwg.org/multipage/images.html#preferred-density-corrected-dimensions)" in the image specification. Another example is `<source>` elements in a `<picture>` element.

The icons in the extension is simpler than images in web pages. For now, we could leave icon-matching to the browser implementation. In the future, if necessary, we can gradually improve the specification.

Reply:
@carlosjeurissen I think #3 makes the most sense, while #1 is what we do today. I would likely do #3 when implementing this for Safari / WebKit.

Reply:
@xeenon mode 3 has the downside of making it less intuitive / easy for authors to add / remove details based on the sizes being used. One may also wonder how a potential density property would play a role here.